File: 00_software_engineering_foundations/29_tictactoe_project/board.rb
class Board
    def initialize(n)
        @grid = Array.new(n) { Array.new(n, '_') }
    end

    def valid?(pos)
        pos.all? { |coordinate| coordinate.between?(0, @grid.length - 1) }
    end

    def empty?(pos)
        row, col = *pos
        @grid[row][col] == '_'
    end

    def place_mark(pos, mark)
        raise 'Invalid position' if !self.valid?(pos)
        raise 'Position not empty' if !self.empty?(pos)

        row, col = *pos
        @grid[row][col] = mark
    end

    def print
        board = ""
        @grid.each do |row|
            row.each do |mark|
                board += "|#{mark}"
            end
            board += "|\n"
        end
        puts board
    end

    def win_row?(mark)
        @grid.any? { |row| row.all? { |row_mark| row_mark == mark } }
    end

    def win_col?(mark)
        @grid.transpose.any? { |col| col.all? { |col_mark| col_mark == mark } }
    end

    def win_diagonal?(mark)
        diagonals = [[], []]

        (0...@grid.length).each do |i|
            diagonals.first << @grid[i][i]
            diagonals.last << @grid[i][-1-i]
        end

        diagonals.any? { |diagonal| diagonal.all? { |diagonal_mark| diagonal_mark == mark } }
    end

    def win?(mark)
        win_row?(mark) || win_col?(mark) || win_diagonal?(mark)
    end

    def empty_positions?
        @grid.any? { |row| row.any? { |mark| mark == "_" } }
    end

    def legal_positions
        positions = []
        (0...@grid.length).each do |row|
            (0...@grid.length).each do |col|
                pos = [row, col]
                positions << pos if empty?(pos)
            end
        end
        positions
    end
end

File: 04_javascript/07_tic_tac_toe_js_edition/board.js
class Board {
  constructor() {
    this.board = [...Array(3)].map(() => [...Array(3)]);
  }

  won() {
    return !!this.winner();
  }

  full() {
    return this.rows().every((row) => row.every((mark) => !!mark));
  }

  valid(pos) {
    return 0 <= pos[0] && pos[0] <= 2 && 0 <= pos[1] && pos[1] <= 2;
  }

  empty(pos) {
    return this.board[pos[0]][pos[1]] === undefined;
  }

  emptyPositions() {
    const positions = [];
    for (let row = 0; row < 3; row++) {
      for (let col = 0; col < 3; col++) {
        const pos = [row, col];
        if (this.empty(pos)) {
          positions.push(pos);
        }
      }
    }
    return positions;
  }

  placeMark(pos, mark) {
    if (this.valid(pos) && this.empty(pos)) {
      this.board[pos[0]][pos[1]] = mark;
      return true;
    }
    return false;
  }

  winner() {
    const winLines = [...this.rows(), ...this.cols(), ...this.diags()];
    for (const line of winLines) {
      if (line.every((mark) => !!mark && mark === line[0])) {
        return line[0];
      }
    }
  }

  rows() {
    return this.board;
  }

  cols() {
    const columns = [[], [], []];
    for (let row = 0; row < 3; row++) {
      for (let col = 0; col < 3; col++) {
        columns[col][row] = this.board[row][col];
      }
    }
    return columns;
  }

  diags() {
    const diagonals = [[], []];
    for (let i = 0; i < 3; i++) {
      diagonals[0][i] = this.board[i][i];
      diagonals[1][i] = this.board[i][this.board.length - 1 - i];
    }
    return diagonals;
  }

  print() {
    this.rows().forEach((row) =>
      console.log(row.map((mark) => (!!mark ? mark : '_')))
    );
  }
}

module.exports = Board;
